### PR TITLE
[#31] Change TOML to use PrefixMap

### LINKED ISSUE #31: Replace 'HashMap' in 'TOML' with our 'PrefixTree'
This involves:

- [x] Chane `TOML` type
- [x] Add `insert` and `lookup` functions for `PrefixMap` (basically just refactoring current implementation)
- [x] Change parser (removing merging, using `insert` from `PrefixMap`)
- [x] Change renderer

### PR DESCRIPTION
**Pull request by vrom911**
Resolves #31

### CODE REVIEW

File: src/Toml/Parser.hs [hunk 1]
```diff
@@ -81,7 +78,7 @@ keyComponentP :: Parser Text
 keyComponentP = bareKeyP <|> (quote <$> stringP)
 
 keyP :: Parser Key
-keyP = Key <$> NC.sepBy1 keyComponentP (char '.')
+keyP = Key <$> NonEmpty.map Piece <$> NC.sepBy1 keyComponentP (char '.')
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: I think it's better to change type of `keyComponentP` to return `Piece`.

File: src/Toml/Parser.hs [hunk 2]
```diff
@@ -122,55 +119,23 @@ keyValP = do
...
+    makeToml :: [(Key, AnyValue)] -> TOML
+    makeToml kv = TOML (HashMap.fromList kv) mempty
 
 tomlP :: Parser TOML
 tomlP = do
     kvs    <- many keyValP
     tables <- many tableHeaderP
     pure TOML { tomlPairs  = HashMap.fromList kvs
-              , tomlTables = merge $ List.sortBy (comparing thName) tables
+              , tomlTables = foldr (\TableHeader{..} -> insert thName thKeys) mempty tables
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: It's usually better to use `foldl'` function. I think you can use `foldl'` and implement this lambda in `where`. Or maybe you can somehow use `fromList` function here by remove `TableHeader` data type and replacing it with just a pair...

File: src/Toml/Parser.hs [hunk 3]
```diff
@@ -122,55 +119,23 @@ keyValP = do
...
-{- Groups by prefix. So the following list:
-
-["a", "a.x", "a.y", "b.1", "b.2", "c"]
-
-will be converted in the following list of non empty lists:
-
-[ "a" :| ["a.x", "a.y"], "b.1" :| [], "b.2" :| [], "c" :| [] ]
-
--}
-groupByPrefix :: [TableHeader] -> [NonEmpty TableHeader]
```
Comment (ChShersh) [inline src/Toml/Parser.hs]: It's so nice to see this ugly code removed!

File: src/Toml/Printer.hs [hunk 4]
```diff
@@ -81,15 +82,29 @@ prettyKeyValue i = Text.concat . map kvText . HashMap.toList
     showText :: Show a => a -> Text
     showText = Text.pack . show
 
-prettyTables :: Int -> HashMap Key TOML -> Text
-prettyTables i = Text.concat . map prettyTable . HashMap.toList
+prettyTables :: Int -> Text -> PrefixMap TOML -> Text
+prettyTables i pref = Text.concat . map prettyTable . HashMap.elems
```
Comment (ChShersh) [inline src/Toml/Printer.hs]: Code in this function looks not very elegant... But I don't see better way to implement it. Need to think more, it's not so obvious...

File: src/Toml/PrefixTree.hs [hunk 5]
```diff
@@ -139,3 +158,9 @@ lookupT lk (Branch pref mv prefMap) =
 -- | Looks up the value at a key in the 'PrefixMap'.
 lookup :: Key -> PrefixMap a -> Maybe a
 lookup k@(p :|| _) prefMap = HashMap.lookup p prefMap >>= lookupT k
+
+fromList :: [(Key, a)] -> PrefixMap a
```
Comment (ChShersh) [inline src/Toml/PrefixTree.hs]: I think this function requires some documentation.

File: src/Toml/Parser.hs (hunk 2 above)
Comment (vrom911) [reply src/Toml/Parser.hs]: That actually cool idea! Will work on it :wink:

**Pull request merged**